### PR TITLE
planner: outer join elimination doesn't consider the ORDER BY items from the agg func

### LINKED ISSUE #18216: SQL Parser Error: cannot resolve column reference with nested group_concat
## Bug Report

Preview:
````
mysql> select group_concat(c order by (select group_concat(c order by a) from t2  where a=t1.a)) from t1;
ERROR 1105 (HY000): Can't find column Column#7 in schema Column: [test.t1.c] Unique key: []
````

### 1. Minimal reproduce step (Required)

````sql
drop table if exists t1, t2;
create table t1 (a int, c int); 
insert into t1 values (1, 1), (1, 2), (2, 3), (2, 4); 
create table t2 (a int, c int); 
insert into t2 values (1, 1), (1, 2), (2, 3), (2, 4); 
select group_concat(c order by (select group_concat(c order by a) from t2 where a=t1.a)) from t1; 
````

### 2. What did you expect to see? (Required)

It should execute without error and produce result similar to the following:
````
-- there are many possible orders regrading the final result
mysql> select group_concat(c order by (select group_concat(c order by a) from t2 where a=t1.a)) from t1;
+-----------------------------------------------------------------------------------+
| group_concat(c order by (select group_concat(c order by a) from t2 where a=t1.a)) |
+-----------------------------------------------------------------------------------+
| 2,1,4,3                                                                           |
+-----------------------------------------------------------------------------------+
1 row in set (0.00 sec)
````
### 3. What did you see instead (Required)
The parser cannot resolve the column name and produces meaningless error message to the user.

### 4. Affected version (Required)

Master branch
````
commit 6fa2badf388445f8f91e72b0aad0c217b7fcfb83 (HEAD -> master, origin/master, origin/HEAD)
Author: cfzjywxk <lsswxr@163.com>
Date:   Wed Jun 24 17:43:10 2020 +0800
````

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #18216 

Problem Summary:

Outer join is wrongly eliminated since we don't take the ORDER BY items from the GROUP_CONCAT.

### What is changed and how it works?

Add it.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that GROUP CONCAT with ORDER BY might fail when the ORDER BY clause contains a correlated subquery.
```

### CODE REVIEW

**Review by AilinKid — APPROVED** [review]
LGTM

**Review by time-and-fate — APPROVED** [review]

File: planner/core/integration_test.go [hunk 1]
```diff
@@ -7675,3 +7675,17 @@ func TestIssue38295(t *testing.T) {
...
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1, t2;")
+	tk.MustExec("create table t1 (a int, c int);")
+	tk.MustExec("insert into t1 values (1, 1), (1, 2), (2, 3), (2, 4);")
+	tk.MustExec("create table t2 (a int, c int);")
+	tk.MustExec("insert into t2 values (1, 1), (1, 2), (2, 3), (2, 4);")
+	// The output might be unstable.
+	tk.MustExec("select group_concat(c order by (select group_concat(c order by a) from t2 where a=t1.a)) from t1; ")
+	tk.MustQuery("select group_concat(c order by (select group_concat(c order by c) from t2 where a=t1.a), c desc) from t1;").Check(testkit.Rows("2,1,4,3"))
```
Comment (hawkingrei) [inline planner/core/integration_test.go:7690]: Will it be unstable if using ```.Sort().Check...``` ?

Comment (winoros) [reply planner/core/integration_test.go:7690]: The output is just one row. The aggregation order is not stable.

**Pull request merged**